File: sdks/dotnet/src/Autlantic.Billing/AutlanticBilling.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Autlantic.Billing;

/// <summary>Merchant server client for https://billing.autlantic.com (hosted only).</summary>
public sealed class AutlanticBilling : IDisposable
{
    private static readonly HashSet<int> RetryableStatus = new()
    {
        408, 429, 500, 502, 503, 504,
    };

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = null,
    };

    private readonly HttpClient _httpClient;
    private readonly bool _ownsHttpClient;

    public string ApiKey { get; }
    public string ApiBaseUrl { get; }
    public string? MerchantId { get; }
    public double TimeoutSec { get; }
    public int MaxRetries { get; }
    public string Mode { get; }

    public AutlanticBilling(string apiKey)
        : this(apiKey, "https://billing.autlantic.com", null, 30.0, 2)
    {
    }

    public AutlanticBilling(string apiKey, string? apiBaseUrl, string? merchantId)
        : this(apiKey, apiBaseUrl, merchantId, 30.0, 2)
    {
    }

    public AutlanticBilling(
        string apiKey,
        string? apiBaseUrl,
        string? merchantId,
        double timeoutSec,
        int maxRetries,
        HttpClient? httpClient = null)
    {
        var key = (apiKey ?? "").Trim();
        if (string.IsNullOrEmpty(key))
        {
            throw new AutlanticBillingException("api_key is required", "configuration");
        }

        ApiKey = key;
        var baseUrl = string.IsNullOrWhiteSpace(apiBaseUrl)
            ? "https://billing.autlantic.com"
            : apiBaseUrl.Trim();
        while (baseUrl.EndsWith('/'))
        {
            baseUrl = baseUrl[..^1];
        }

        ApiBaseUrl = baseUrl;
        MerchantId = string.IsNullOrWhiteSpace(merchantId) ? null : merchantId.Trim();
        TimeoutSec = timeoutSec;
        MaxRetries = Math.Max(0, maxRetries);
        Mode = BillingModeFromApiKey(ApiKey);

        if (httpClient is not null)
        {
            _httpClient = httpClient;
            _ownsHttpClient = false;
        }
        else
        {
            _httpClient = new HttpClient
            {
                Timeout = TimeSpan.FromMilliseconds(Math.Max(1, (long)(timeoutSec * 1000))),
            };
            _ownsHttpClient = true;
        }
    }

    /// <summary>Build a client from environment variables.</summary>
    public static AutlanticBilling FromEnv()
        => FromEnv(Environment.GetEnvironmentVariables()
            .Cast<System.Collections.DictionaryEntry>()
            .ToDictionary(
                e => e.Key?.ToString() ?? "",
                e => e.Value?.ToString()));

    public static AutlanticBilling FromEnv(IReadOnlyDictionary<string, string?> env)
    {
        var apiKey = TrimOrEmpty(GetEnv(env, "AUTLANTIC_BILLING_API_KEY"));
        if (string.IsNullOrEmpty(apiKey))
        {
            throw new AutlanticBillingException(
                "AUTLANTIC_BILLING_API_KEY is required",
                "configuration");
        }

        var baseUrl = TrimOrEmpty(GetEnv(env, "AUTLANTIC_BILLING_API_URL"));
        if (string.IsNullOrEmpty(baseUrl))
        {
            baseUrl = "https://billing.autlantic.com";
        }

        var merchant = TrimOrEmpty(GetEnv(env, "AUTLANTIC_BILLING_MERCHANT_ID"));
        return new AutlanticBilling(apiKey, baseUrl, string.IsNullOrEmpty(merchant) ? null : merchant);
    }

    public static string BillingModeFromApiKey(string? apiKey)
    {
        var key = (apiKey ?? "").Trim();
        return key.Contains("_live_", StringComparison.Ordinal) ? "live" : "test";
    }

    public Task<JsonObject> ListProductsAsync(CancellationToken cancellationToken = default)
        => RequestAsync("GET", "/v1/products", null, cancellationToken);

    public JsonObject ListProducts()
        => ListProductsAsync().GetAwaiter().GetResult();

    public Task<JsonObject> CreateSubscriptionAsync(
        JsonNode? body,
        CancellationToken cancellationToken = default)
        => RequestAsync("POST", "/v1/subscriptions", body, cancellationToken);

    public JsonObject CreateSubscription(JsonNode? body)
        => CreateSubscriptionAsync(body).GetAwaiter().GetResult();

    public JsonObject CreateSubscription(IDictionary<string, object?> body)
        => CreateSubscription(ToJsonObject(body));

    public Task<JsonObject> GetSubscriptionAsync(
        string subscriptionId,
        CancellationToken cancellationToken = default)
        => RequestAsync("GET", $"/v1/subscriptions/{Enc(subscriptionId)}", null, cancellationToken);

    public JsonObject GetSubscription(string subscriptionId)
        => GetSubscriptionAsync(subscriptionId).GetAwaiter().GetResult();

    public Task<JsonObject> ActivateSubscriptionAsync(
        string subscriptionId,
        CancellationToken cancellationToken = default)
        => RequestAsync(
            "POST",
            $"/v1/subscriptions/{Enc(subscriptionId)}/activate",
            new JsonObject(),
            cancellationToken);

    public JsonObject ActivateSubscription(string subscriptionId)
        => ActivateSubscriptionAsync(subscriptionId).GetAwaiter().GetResult();

    public Task<JsonObject> CancelSubscriptionAsync(
        string subscriptionId,
        JsonNode? body = null,
        CancellationToken cancellationToken = default)
        => RequestAsync(
            "POST",
            $"/v1/subscriptions/{Enc(subscriptionId)}/cancel",
            body ?? new JsonObject(),
            cancellationToken);

    public JsonObject CancelSubscription(string subscriptionId, JsonNode? body = null)
        => CancelSubscriptionAsync(subscriptionId, body).GetAwaiter().GetResult();

    public JsonObject CancelSubscription(string subscriptionId, IDictionary<string, object?>? body)
        => CancelSubscription(subscriptionId, body is null ? null : ToJsonObject(body));

    public Task<JsonObject> CreatePaymentAsync(
        JsonNode? body,
        CancellationToken cancellationToken = default)
        => RequestAsync("POST", "/v1/payments", body, cancellationToken);

    public JsonObject CreatePayment(JsonNode? body)
        => CreatePaymentAsync(body).GetAwaiter().GetResult();

    public JsonObject CreatePayment(IDictionary<string, object?> body)
        => CreatePayment(ToJsonObject(body));

    public Task<JsonObject> GetPaymentAsync(
        string paymentId,
        CancellationToken cancellationToken = default)
        => RequestAsync("GET", $"/v1/payments/{Enc(paymentId)}", null, cancellationToken);

    public JsonObject GetPayment(string paymentId)
        => GetPaymentAsync(paymentId).GetAwaiter().GetResult();

    public Task<JsonObject> CreatePaymentLinkAsync(
        JsonNode? body,
        CancellationToken cancellationToken = default)
        => RequestAsync("POST", "/v1/payment-links", body, cancellationToken);

    public JsonObject CreatePaymentLink(JsonNode? body)
        => CreatePaymentLinkAsync(body).GetAwaiter().GetResult();

    public JsonObject CreatePaymentLink(IDictionary<string, object?> body)
        => CreatePaymentLink(ToJsonObject(body));

    public Task<JsonObject> ListPaymentLinksAsync(CancellationToken cancellationToken = default)
        => RequestAsync("GET", "/v1/payment-links", null, cancellationToken);

    public JsonObject ListPaymentLinks()
        => ListPaymentLinksAsync().GetAwaiter().GetResult();

    public Task<JsonObject> GetPaymentLinkAsync(
        string linkId,
        CancellationToken cancellationToken = default)
        => RequestAsync("GET", $"/v1/payment-links/{Enc(linkId)}", null, cancellationToken);

    public JsonObject GetPaymentLink(string linkId)
        => GetPaymentLinkAsync(linkId).GetAwaiter().GetResult();

    public Task<JsonObject> DisablePaymentLinkAsync(
        string linkId,
        CancellationToken cancellationToken = default)
        => RequestAsync(
            "POST",
            $"/v1/payment-links/{Enc(linkId)}/disable",
            new JsonObject(),
            cancellationToken);

    public JsonObject DisablePaymentLink(string linkId)
        => DisablePaymentLinkAsync(linkId).GetAwaiter().GetResult();

    public Task<JsonObject> ListInvoicesAsync(
        string? subscriptionId = null,
        CancellationToken cancellationToken = default)
    {
        var path = "/v1/invoices";
        if (!string.IsNullOrEmpty(subscriptionId))
        {
            path += $"?subscriptionId={Enc(subscriptionId)}";
        }

        return RequestAsync("GET", path, null, cancellationToken);
    }

    public JsonObject ListInvoices(string? subscriptionId = null)
        => ListInvoicesAsync(subscriptionId).GetAwaiter().GetResult();

    private async Task<JsonObject> RequestAsync(
        string method,
        string path,
        JsonNode? body,
        CancellationToken cancellationToken)
    {
        var requestId =
            $"req_{DateTimeOffset.UtcNow.ToUnixTimeMilliseconds():x}_{RandomHex(4)}";
        var url = ApiBaseUrl + path;
        string? payload = null;
        if (body is not null)
        {
            payload = body.ToJsonString(JsonOptions);
        }

        string? idemKey = null;
        if (string.Equals(method, "POST", StringComparison.Ordinal))
        {
            idemKey =
                $"sdk_{DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()}_{RandomHex(4)}";
        }

        var attempts = MaxRetries + 1;
        AutlanticBillingException? lastError = null;

        for (var attempt = 0; attempt < attempts; attempt++)
        {
            try
            {
                using var request = new HttpRequestMessage(new HttpMethod(method), url);
                request.Headers.TryAddWithoutValidation("Accept", "application/json");
                request.Headers.TryAddWithoutValidation("User-Agent", Version.UserAgent);
                request.Headers.TryAddWithoutValidation("X-Autlantic-Api-Key", ApiKey);
                request.Headers.TryAddWithoutValidation(
                    "X-Autlantic-Sdk-Version",
                    Version.SdkVersion);
                request.Headers.TryAddWithoutValidation(
                    "X-Autlantic-Client-Request-Id",
                    requestId);
                request.Headers.TryAddWithoutValidation(
                    "Autlantic-Version",
                    Version.AutlanticApiVersion);

                if (idemKey is not null)
                {
                    request.Headers.TryAddWithoutValidation("Idempotency-Key", idemKey);
                }

                if (string.Equals(method, "POST", StringComparison.Ordinal)
                    || string.Equals(method, "PUT", StringComparison.Ordinal)
                    || string.Equals(method, "PATCH", StringComparison.Ordinal))
                {
                    var content = payload ?? "{}";
                    request.Content = new StringContent(content, Encoding.UTF8, "application/json");
                }
                else if (payload is not null)
                {
                    request.Content = new StringContent(payload, Encoding.UTF8, "application/json");
                }

                using var response = await _httpClient
                    .SendAsync(request, cancellationToken)
                    .ConfigureAwait(false);
                var status = (int)response.StatusCode;
                var responseBody = await response.Content
                    .ReadAsStringAsync(cancellationToken)
                    .ConfigureAwait(false);

                JsonNode? decoded;
                if (string.IsNullOrEmpty(responseBody))
                {
                    decoded = new JsonObject();
                }
                else
                {
                    try
                    {
                        decoded = JsonNode.Parse(responseBody);
                    }
                    catch (JsonException)
                    {
                        decoded = JsonValue.Create(responseBody);
                    }
                }

                if (status is >= 200 and < 300)
                {
                    if (decoded is not JsonObject obj)
                    {
                        throw new AutlanticBillingException(
                            "Unexpected response shape",
                            null,
                            status,
                            requestId,
                            decoded);
                    }

                    return obj;
                }

                var message = $"HTTP {status}";
                string? code = null;
                if (decoded is JsonObject map)
                {
                    if (map["error"] is not null)
                    {
                        message = map["error"]!.ToString();
                    }

                    if (map["code"] is JsonValue codeVal
                        && codeVal.TryGetValue<string>(out var codeStr))
                    {
                        code = codeStr;
                    }
                }

                lastError = new AutlanticBillingException(
                    message,
                    code,
                    status,
                    requestId,
                    decoded);

                if (RetryableStatus.Contains(status) && attempt + 1 < attempts)
                {
                    await SleepBackoffAsync(attempt, cancellationToken).ConfigureAwait(false);
                    continue;
                }

                throw lastError;
            }
            catch (AutlanticBillingException)
            {
                throw;
            }
            catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
            {
                throw;
            }
            catch (Exception e)
            {
                lastError = new AutlanticBillingException(
                    "Network error: " + e.Message,
                    "network_error",
                    null,
                    requestId);
                if (attempt + 1 < attempts)
                {
                    await SleepBackoffAsync(attempt, cancellationToken).ConfigureAwait(false);
                    continue;
                }

                throw lastError;
            }
        }

        throw lastError
              ?? new AutlanticBillingException(
                  "Request failed",
                  "network_error",
                  null,
                  requestId);
    }

    public void Dispose()
    {
        if (_ownsHttpClient)
        {
            _httpClient.Dispose();
        }
    }

    private static async Task SleepBackoffAsync(int attempt, CancellationToken cancellationToken)
    {
        var ms = (int)(250 * Math.Pow(2, attempt));
        await Task.Delay(ms, cancellationToken).ConfigureAwait(false);
    }

    private static string Enc(string value)
        => Uri.EscapeDataString(value);

    private static string TrimOrEmpty(string? value)
        => value?.Trim() ?? "";

    private static string? GetEnv(IReadOnlyDictionary<string, string?> env, string key)
        => env.TryGetValue(key, out var value) ? value : null;

    private static string RandomHex(int numBytes)
    {
        var bytes = RandomNumberGenerator.GetBytes(numBytes);
        return Convert.ToHexString(bytes).ToLowerInvariant();
    }

    private static JsonObject ToJsonObject(IDictionary<string, object?> body)
    {
        var obj = new JsonObject();
        foreach (var (key, value) in body)
        {
            obj[key] = value is null ? null : JsonSerializer.SerializeToNode(value);
        }

        return obj;
    }
}

File: sdks/java/src/main/java/com/autlantic/billing/AutlanticBilling.java
package com.autlantic.billing;

import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.security.SecureRandom;
import java.time.Duration;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Set;
import org.json.JSONObject;

/**
 * Merchant server client for https://billing.autlantic.com (hosted only).
 */
public final class AutlanticBilling {
  private static final Set<Integer> RETRYABLE_STATUS =
      Set.of(408, 429, 500, 502, 503, 504);
  private static final SecureRandom RANDOM = new SecureRandom();

  private final String apiKey;
  private final String apiBaseUrl;
  private final String merchantId;
  private final double timeoutSec;
  private final int maxRetries;
  private final String mode;
  private final HttpClient httpClient;

  public AutlanticBilling(String apiKey) {
    this(apiKey, "https://billing.autlantic.com", null, 30.0, 2);
  }

  public AutlanticBilling(String apiKey, String apiBaseUrl, String merchantId) {
    this(apiKey, apiBaseUrl, merchantId, 30.0, 2);
  }

  public AutlanticBilling(
      String apiKey,
      String apiBaseUrl,
      String merchantId,
      double timeoutSec,
      int maxRetries) {
    String key = apiKey == null ? "" : apiKey.trim();
    if (key.isEmpty()) {
      throw new AutlanticBillingException("api_key is required", "configuration");
    }
    this.apiKey = key;
    String base =
        apiBaseUrl == null || apiBaseUrl.trim().isEmpty()
            ? "https://billing.autlantic.com"
            : apiBaseUrl.trim();
    while (base.endsWith("/")) {
      base = base.substring(0, base.length() - 1);
    }
    this.apiBaseUrl = base;
    this.merchantId =
        merchantId == null || merchantId.trim().isEmpty() ? null : merchantId.trim();
    this.timeoutSec = timeoutSec;
    this.maxRetries = Math.max(0, maxRetries);
    this.mode = billingModeFromApiKey(this.apiKey);
    this.httpClient =
        HttpClient.newBuilder()
            .connectTimeout(Duration.ofMillis(Math.max(1, (long) (timeoutSec * 1000))))
            .build();
  }

  public String getApiKey() {
    return apiKey;
  }

  public String getApiBaseUrl() {
    return apiBaseUrl;
  }

  public String getMerchantId() {
    return merchantId;
  }

  public double getTimeoutSec() {
    return timeoutSec;
  }

  public int getMaxRetries() {
    return maxRetries;
  }

  public String getMode() {
    return mode;
  }

  /** Build a client from environment variables. */
  public static AutlanticBilling fromEnv() {
    return fromEnv(System.getenv());
  }

  public static AutlanticBilling fromEnv(Map<String, String> env) {
    Map<String, String> e = env != null ? env : Map.of();
    String apiKey = trimOrEmpty(e.get("AUTLANTIC_BILLING_API_KEY"));
    if (apiKey.isEmpty()) {
      throw new AutlanticBillingException(
          "AUTLANTIC_BILLING_API_KEY is required", "configuration");
    }
    String base = trimOrEmpty(e.get("AUTLANTIC_BILLING_API_URL"));
    if (base.isEmpty()) {
      base = "https://billing.autlantic.com";
    }
    String merchant = trimOrEmpty(e.get("AUTLANTIC_BILLING_MERCHANT_ID"));
    return new AutlanticBilling(apiKey, base, merchant.isEmpty() ? null : merchant);
  }

  public static String billingModeFromApiKey(String apiKey) {
    String key = apiKey == null ? "" : apiKey.trim();
    return key.contains("_live_") ? "live" : "test";
  }

  @SuppressWarnings("unchecked")
  public Map<String, Object> listProducts() {
    return request("GET", "/v1/products", null);
  }

  public Map<String, Object> createSubscription(Map<String, Object> body) {
    return request("POST", "/v1/subscriptions", body);
  }

  public Map<String, Object> getSubscription(String subscriptionId) {
    return request("GET", "/v1/subscriptions/" + enc(subscriptionId), null);
  }

  public Map<String, Object> activateSubscription(String subscriptionId) {
    return request(
        "POST", "/v1/subscriptions/" + enc(subscriptionId) + "/activate", Map.of());
  }

  public Map<String, Object> cancelSubscription(String subscriptionId) {
    return cancelSubscription(subscriptionId, null);
  }

  public Map<String, Object> cancelSubscription(
      String subscriptionId, Map<String, Object> body) {
    return request(
        "POST",
        "/v1/subscriptions/" + enc(subscriptionId) + "/cancel",
        body != null ? body : Map.of());
  }

  public Map<String, Object> createPayment(Map<String, Object> body) {
    return request("POST", "/v1/payments", body);
  }

  public Map<String, Object> getPayment(String paymentId) {
    return request("GET", "/v1/payments/" + enc(paymentId), null);
  }

  public Map<String, Object> createPaymentLink(Map<String, Object> body) {
    return request("POST", "/v1/payment-links", body);
  }

  public Map<String, Object> listPaymentLinks() {
    return request("GET", "/v1/payment-links", null);
  }

  public Map<String, Object> getPaymentLink(String linkId) {
    return request("GET", "/v1/payment-links/" + enc(linkId), null);
  }

  public Map<String, Object> disablePaymentLink(String linkId) {
    return request("POST", "/v1/payment-links/" + enc(linkId) + "/disable", Map.of());
  }

  public Map<String, Object> listInvoices() {
    return listInvoices(null);
  }

  public Map<String, Object> listInvoices(String subscriptionId) {
    String path = "/v1/invoices";
    if (subscriptionId != null && !subscriptionId.isEmpty()) {
      path += "?subscriptionId=" + enc(subscriptionId);
    }
    return request("GET", path, null);
  }

  @SuppressWarnings("unchecked")
  private Map<String, Object> request(
      String method, String path, Map<String, Object> body) {
    String requestId =
        "req_"
            + Long.toHexString(System.currentTimeMillis())
            + "_"
            + randomHex(4);
    String url = apiBaseUrl + path;
    String payload = body == null ? null : new JSONObject(body).toString();
    String idemKey = null;
    if ("POST".equals(method)) {
      idemKey = "sdk_" + System.currentTimeMillis() + "_" + randomHex(4);
    }

    int attempts = maxRetries + 1;
    AutlanticBillingException lastError = null;

    for (int attempt = 0; attempt < attempts; attempt++) {
      try {
        HttpRequest.Builder builder =
            HttpRequest.newBuilder()
                .uri(URI.create(url))
                .timeout(Duration.ofMillis(Math.max(1, (long) (timeoutSec * 1000))))
                .header("Accept", "application/json")
                .header("User-Agent", Version.USER_AGENT)
                .header("X-Autlantic-Api-Key", apiKey)
                .header("X-Autlantic-Sdk-Version", Version.SDK_VERSION)
                .header("X-Autlantic-Client-Request-Id", requestId)
                .header("Autlantic-Version", Version.AUTLANTIC_API_VERSION);

        if (payload != null) {
          builder.header("Content-Type", "application/json");
        }
        if (idemKey != null) {
          builder.header("Idempotency-Key", idemKey);
        }

        if ("GET".equals(method)) {
          builder.GET();
        } else if ("POST".equals(method)) {
          builder.POST(
              HttpRequest.BodyPublishers.ofString(
                  payload != null ? payload : "{}", StandardCharsets.UTF_8));
        } else {
          builder.method(
              method,
              payload != null
                  ? HttpRequest.BodyPublishers.ofString(payload, StandardCharsets.UTF_8)
                  : HttpRequest.BodyPublishers.noBody());
        }

        HttpResponse<String> response =
            httpClient.send(builder.build(), HttpResponse.BodyHandlers.ofString());
        int status = response.statusCode();
        String responseBody = response.body() == null ? "" : response.body();

        Object decoded;
        if (responseBody.isEmpty()) {
          decoded = new LinkedHashMap<String, Object>();
        } else {
          try {
            decoded = new JSONObject(responseBody).toMap();
          } catch (Exception e) {
            decoded = responseBody;
          }
        }

        if (status >= 200 && status < 300) {
          if (!(decoded instanceof Map)) {
            throw new AutlanticBillingException(
                "Unexpected response shape", null, status, requestId, decoded);
          }
          return (Map<String, Object>) decoded;
        }

        String message = "HTTP " + status;
        String code = null;
        if (decoded instanceof Map<?, ?> map) {
          Object err = map.get("error");
          if (err != null) {
            message = String.valueOf(err);
          }
          Object c = map.get("code");
          if (c instanceof String) {
            code = (String) c;
          }
        }
        lastError =
            new AutlanticBillingException(message, code, status, requestId, decoded);
        if (RETRYABLE_STATUS.contains(status) && attempt + 1 < attempts) {
          sleepBackoff(attempt);
          continue;
        }
        throw lastError;
      } catch (AutlanticBillingException e) {
        throw e;
      } catch (InterruptedException e) {
        Thread.currentThread().interrupt();
        throw new AutlanticBillingException(
            "Network error: interrupted", "network_error", null, requestId);
      } catch (Exception e) {
        lastError =
            new AutlanticBillingException(
                "Network error: " + e.getMessage(),
                "network_error",
                null,
                requestId);
        if (attempt + 1 < attempts) {
          sleepBackoff(attempt);
          continue;
        }
        throw lastError;
      }
    }

    if (lastError != null) {
      throw lastError;
    }
    throw new AutlanticBillingException("Request failed", "network_error", null, requestId);
  }

  private static void sleepBackoff(int attempt) {
    try {
      Thread.sleep((long) (250 * Math.pow(2, attempt)));
    } catch (InterruptedException e) {
      Thread.currentThread().interrupt();
    }
  }

  private static String enc(String value) {
    return URLEncoder.encode(value, StandardCharsets.UTF_8).replace("+", "%20");
  }

  private static String trimOrEmpty(String value) {
    return value == null ? "" : value.trim();
  }

  private static String randomHex(int numBytes) {
    byte[] bytes = new byte[numBytes];
    RANDOM.nextBytes(bytes);
    StringBuilder sb = new StringBuilder(numBytes * 2);
    for (byte b : bytes) {
      sb.append(String.format("%02x", b));
    }
    return sb.toString();
  }
}
